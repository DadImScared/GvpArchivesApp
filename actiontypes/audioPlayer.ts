export const SHOW_PLAYER = "audio/SHOW_PLAYER";
export const TOGGLE_PLAYING = "audio/TOGGLE_PLAYING";
export const SEEK_TO = "audio/SEEK_TO";

// playlist types
export const SET_PLAYLIST = "audio/SET_PLAYLIST";
export const UPDATE_CURRENT_INDEX = "audio/UPDATE_CURRENT_INDEX";

// single song types
export const SET_SONG = "audio/SET_SONG";
export const SET_PLAYING = "audio/SET_PLAYING";

// expo player duration update

export const UPDATE_PLAYBACKINSTANCE = "audio/UPDATE_PLAYBACKINSTANCE";
