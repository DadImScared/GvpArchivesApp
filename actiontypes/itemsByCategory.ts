export const GET_ITEMS_BY_CATEGORY = "itemsByCategory/GET_ITEMS_BY_CATEGORY";
export const ADD_ITEMS_BY_CATEGORY = "itemsByCategory/ADD_ITEMS_BY_CATEGORY";
export const UPDATE_ITEMS_BY_CATEGORY = "itemsByCategory/UPDATE_ITEMS";
