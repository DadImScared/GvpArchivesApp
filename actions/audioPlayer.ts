import { action } from "typesafe-actions";

import * as AudioActionTypes from "../actiontypes/audioPlayer";

export const togglePlaying = () => action(AudioActionTypes.TOGGLE_PLAYING);

export const setPlaying = (playing: boolean) => action(AudioActionTypes.SET_PLAYING, { playing });

export const showPlayer = (shouldShowPlayer: boolean) => {
  return action(AudioActionTypes.SHOW_PLAYER, { showPlayer: shouldShowPlayer });
};

export const setPlaylist = (name: string, items: string[], currentIndex = 0) => {
  return action(
    AudioActionTypes.SET_PLAYLIST,
    {
      currentIndex,
      items,
      name,
      playerType: "playlist" as any,
      playing: true,
      showPlayer: true,
    }
  );
};

export const updateIndex = (newIndex: number) => action(AudioActionTypes.UPDATE_CURRENT_INDEX, { newIndex });

export const setSong = (songUrl: string, songName: string) => {
  return action(
    AudioActionTypes.SET_SONG,
    {
      songName,
      songUrl,
      playerType: "song",
      playing: true,
      showPlayer: true
    }
  );
};

export interface IAudioPlayerActions {
  updateIndex: typeof updateIndex;
  togglePlaying: typeof togglePlaying;
  setSong: typeof setSong;
  setPlaying: typeof setPlaying;
  setPlaylist: typeof setPlaylist;
  showPlayer: typeof showPlayer;
}

export default {
  setPlaying,
  setPlaylist,
  setSong,
  showPlayer,
  togglePlaying,
  updateIndex,
};
