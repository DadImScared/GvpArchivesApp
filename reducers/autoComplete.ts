import moment from "moment";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import autoComplete from "../actions/autoComplete";
import * as ActionTypes from "../actiontypes/autoComplete";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "./index";
import { ILoadingState } from "./loading";
import { getQueryId, getSearch, ISearchState } from "./search";

export interface IAutoCompleteState {
  [key: string]: {
    results: string[];
    timeToLive: number;
  };
}

export const getInitialAutoCompleteState = () => ({});

export default function(state = getInitialAutoCompleteState(), action: ActionType<typeof autoComplete>) {
  switch (action.type) {
    case ActionTypes.ADD_AUTOCOMPLETE:
      const { payload: { query, categories, results } } = action;
      return {
        ...state,
        [getQueryId(query, categories)]: {
          results,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    default:
      return state;
  }
}

interface IGetAutoCompleteReturn {
  results: string[];
  timeToLive: number;
}

export interface IAutoCompleteScreenData {
  autoComplete: IGetAutoCompleteReturn;
  query: string;
  queryId: string;
  categories: string[];
  isLoading: boolean;
}

export const getAutoCompleteScreen = createSelector<
  IReducerState,
  IAutoCompleteState,
  ISearchState,
  ILoadingState,
  IAutoCompleteScreenData
>(
  [(state) => state.autoComplete, getSearch, (state) => state.loading],
  (autoCompleteState, { categories, query }, loadingState) => {
    const queryId = getQueryId(query, categories);
    return {
      categories,
      query,
      queryId,
      autoComplete: autoCompleteState[queryId] || { results: [], timeToLive: 0 },
      isLoading: loadingState[`${AUTOCOMPLETE_QUERY}${queryId}`]
    };
  }
);
