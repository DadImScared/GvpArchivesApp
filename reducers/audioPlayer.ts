import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import * as actions from "../actions";
import * as AudioActionTypes from "../actiontypes/audioPlayer";
import { IReducerState } from "./index";
import { getItemsById, ItemsByIdState } from "./itemsById";

export enum ShowPlayerStatus {
  OPEN = "OPEN",
  CLOSED = "CLOSED",
  MINIMIZED = "MINIMIZED"
}

export interface IAudioPlayerState {
  currentIndex: number;
  currentSongName: string;
  currentSongUrl: string;
  items: string[];
  name: string;
  duration: number;
  position: number;
  playerType: "song" | "playlist";
  playing: boolean;
  seekTo: number;
  showPlayer: ShowPlayerStatus;
  sliderValue: number;
}

export type AudioPlayerStateAndActions<
  T extends keyof IAudioPlayerState | keyof typeof actions.audioPlayer
> = Pick<IAudioPlayerState & typeof actions.audioPlayer, T>;

export const getInitialAudioPlayerState = (): IAudioPlayerState => ({
  playerType: "song",
  playing: false,
  seekTo: 0,
  showPlayer: ShowPlayerStatus.CLOSED,
  sliderValue: 0,

  // values used when audio is playing a playlist
  currentIndex: 0,
  items: [],
  name: "",

  // values used when audio player is playing single song
  currentSongName: "",
  currentSongUrl: "",

  // playback status from expo player
  duration: 0,
  position: 0
});

const initialAudioPlayerState = getInitialAudioPlayerState();

export default function audioPlayer(
  state: IAudioPlayerState = initialAudioPlayerState,
  action: ActionType<typeof actions.audioPlayer>
) {
  switch (action.type) {
    case AudioActionTypes.UPDATE_PLAYBACKINSTANCE:
      const { payload } = action;
      if (payload.isLoaded) {
        return {
          ...state,
          duration: payload.durationMillis || 0,
          playing: payload.didJustFinish ? false : state.playing,
          position: payload.positionMillis || 0,
          sliderValue: payload.positionMillis || 0
        };
      }
      return state;
    case AudioActionTypes.SEEK_END:
      return {
        ...state,
        seekTo: action.payload.seekTo * (state.duration || 0)
      };
    case AudioActionTypes.SET_SHOW_PLAYER:
      return {
        ...state,
        showPlayer: action.payload.status
      };
    case AudioActionTypes.SEEK_TO:
      return {
        ...state,
        sliderValue: action.payload.seekTo * (state.duration || 0)
      };
    case AudioActionTypes.TOGGLE_PLAYING:
      return {
        ...state,
        playing: !state.playing
      };
    case AudioActionTypes.SET_PLAYING:
      return {
        ...state,
        playing: action.payload.playing
      };
    case AudioActionTypes.SET_PLAYLIST:
      return {
        ...state,
        currentIndex: action.payload.currentIndex,
        items: action.payload.items,
        name: action.payload.name,
        playerType: action.payload.playerType,
        playing: action.payload.playing,
        showPlayer: action.payload.showPlayer
      };
    case AudioActionTypes.UPDATE_CURRENT_INDEX:
      return {
        ...state,
        currentIndex: action.payload.newIndex
      };
    case AudioActionTypes.SET_SONG:
      return {
        ...state,
        currentSongName: action.payload.songName,
        currentSongUrl: action.payload.songUrl,
        playerType: action.payload.playerType,
        playing: action.payload.playing,
        showPlayer: state.showPlayer === ShowPlayerStatus.MINIMIZED ? state.showPlayer : action.payload.showPlayer
      };
    default:
      return state;
  }
}

const padWithZero = (time: number) => {
  if (time < 10) {
    return `0${time}`;
  }
  return time.toString();
};

const getTimestamp = (millis: number): string => {
  const totalSeconds = millis / 1000;
  const seconds = Math.floor(totalSeconds % 60);
  const minutes = Math.floor(totalSeconds / 60);

  let timeFormat = `${padWithZero(minutes)}:${padWithZero(seconds)}`;
  if (minutes > 59) {
    const hours = Math.floor(minutes / 60);
    timeFormat = `${hours}:${timeFormat}`;
  }
  return timeFormat;
};

const getAudioPlayerState = (state: IReducerState) => state.audioPlayer;

const getItems = (itemIds: string[], itemsById: ItemsByIdState) => itemIds.map((itemId) => itemsById[itemId]);

export const getButtonGroupData = createSelector(
  [getAudioPlayerState],
  ({ playing }) => ({ playing })
);

export const getTitleData = createSelector(
  [getAudioPlayerState, getItemsById],
  (audio, itemsById) => {
    const { playerType, currentSongName, currentIndex, items: itemIds } = audio;
    const items = getItems(itemIds, itemsById);
    return {
      songName: playerType === "song" ? currentSongName : items[currentIndex].title
    };
  }
);

export const getSeekerData = createSelector(
  [getAudioPlayerState],
  ({ playing, position, duration, sliderValue }) => {
    return ({
      playing,
      sliderPosition: sliderValue / duration,
      timestamp: `${getTimestamp(sliderValue)} / ${getTimestamp(duration)}`
    });
  }
);

export const getAudioPlayerData = createSelector(
  [getAudioPlayerState, getItemsById],
  (
    { playing, items: itemIds, currentSongUrl, currentIndex, playerType, seekTo },
    itemsById
  ) => {
    const items = getItems(itemIds, itemsById);
    return {
      playing,
      seekTo,
      uri: playerType === "song" ? currentSongUrl : items[currentIndex].link
    };
  }
);

export const getShowAudioPlayer = createSelector(
  [getAudioPlayerState],
  ({ showPlayer }) => ({ showPlayer })
);
