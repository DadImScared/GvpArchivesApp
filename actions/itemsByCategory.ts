import axios from "axios";
import { normalize } from "normalizr";
import { Action, ActionCreator } from "redux";
import { batchActions } from "redux-batched-actions";
import { ThunkAction } from "redux-thunk";
import { action } from "typesafe-actions";
import { ADD_ITEMS_BY_CATEGORY, UPDATE_ITEMS_BY_CATEGORY } from "../actiontypes/itemsByCategory";
import { ITEMS_BY_CATEGORY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { itemSchema } from "../schemas/items";
import { addItemsById, BASE_URL } from "./itemsById";
import { loadingEnd, loadingStart } from "./loading";

export type ActionCreatorThunk = ActionCreator<ThunkAction<void, IReducerState, undefined, Action<any>>>;

export const getItemsByCategory: ActionCreatorThunk = (category: string, nextPage?: string | boolean) => {
  return async (dispatch): Promise<void> => {
    const LOADING_ID = `${ITEMS_BY_CATEGORY}${category}`;
    dispatch(loadingStart(LOADING_ID));
    const firstPage = `${BASE_URL}/api/v1/items/${category}`;
    const url = typeof nextPage === "string" ? `${BASE_URL}${nextPage}` : firstPage;
    const response = await axios.get(url);
    if (response) {
      // dispatch batch actions
      const { data } = response;
      const { result: itemIds, entities: { items } } = normalize(data.results, itemSchema);
      dispatch(batchActions([
        addItemsById(items),
        (nextPage ? updateItemsByCategory : addItemsByCategory)({category, itemIds, nextPage: data.nextPage }),
        loadingEnd(LOADING_ID)
      ]));
    }
  };
};

interface ItemsByCategory {
  category: string;
  itemIds: string[];
  nextPage: boolean | string;
}

export const addItemsByCategory  = (data: ItemsByCategory) => {
  return action(ADD_ITEMS_BY_CATEGORY, data);
};

export const updateItemsByCategory = (data: ItemsByCategory) => {
  return action(UPDATE_ITEMS_BY_CATEGORY, data);
};

export interface ItemsByCategoryActions {
  addItemsByCategory: typeof addItemsByCategory;
  updateItemsByCategory: typeof updateItemsByCategory;
}

export default {
  addItemsByCategory,
  updateItemsByCategory
};
