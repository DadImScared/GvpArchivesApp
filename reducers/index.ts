import { combineReducers } from "redux";

import audioPlayer, { IAudioPlayerState } from "./audioPlayer";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
}

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  itemsByCategory,
  itemsById,
  loading,
});

export default reducer;
