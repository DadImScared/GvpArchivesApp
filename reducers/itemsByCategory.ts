import _ from "lodash";
import moment from "moment";
import { NavigationScreenProps } from "react-navigation";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import actions from "../actions";
import { ADD_ITEMS_BY_CATEGORY, UPDATE_ITEMS_BY_CATEGORY } from "../actiontypes/itemsByCategory";
import { ITEMS_BY_CATEGORY } from "../loadingPrefix";
import { IReducerState } from "./index";
import { getItemsById, Item } from "./itemsById";
import { getLoading } from "./loading";

interface ItemsByCategory {
  results: string[];
  nextPage?: boolean | string;
  timeToLive?: number;
}

export interface ItemsByCategoryData {
  results: Item[];
  isLoading: boolean;
  nextPage: boolean | string;
  timeToLive: number;
}

export interface ItemsByCategoryState {
  [key: string]: ItemsByCategory;
}

function itemsByCategory(
    state: ItemsByCategoryState = {},
    action: ActionType<typeof actions.itemsByCategory>
) {
  switch (action.type) {
    case ADD_ITEMS_BY_CATEGORY:
      return {
        ...state,
        [action.payload.category]: {
          nextPage: action.payload.nextPage,
          results: action.payload.itemIds,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    case UPDATE_ITEMS_BY_CATEGORY:
      return {
        ...state,
        [action.payload.category]: {
          ...state[action.payload.category],
          nextPage: action.payload.nextPage,
          results: _.union(state[action.payload.category].results, action.payload.itemIds)
        }
      };
    default:
      return state;
  }
}

export const getItems = (state: IReducerState) => state.itemsByCategory;

export const getCategory = (_state: any, props: NavigationScreenProps) => props.navigation.getParam("category");

export const getItemsByCategoryState = createSelector(
    [getItems, getCategory],
    (items, category) => items[category] || {}
);

export const getItemsByCategory = () => {
  return createSelector(
      [
        getItemsByCategoryState,
        getItemsById,
        getLoading(ITEMS_BY_CATEGORY, "navigation.state.params.category")
      ],
      ({ results = [], nextPage = false, timeToLive = 0 }, items, isLoading): ItemsByCategoryData => ({
        isLoading,
        nextPage,
        timeToLive,
        results: results.map((itemId) => items[itemId])
      })
  );
};

export default itemsByCategory;
