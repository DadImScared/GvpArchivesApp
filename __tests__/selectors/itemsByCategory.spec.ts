import { getInitialAudioPlayerState } from "../../reducers/audioPlayer";
import { getItemsByCategory } from "../../reducers/itemsByCategory";

describe("itemsByCategory selector", () => {
  it("should getItemsByCategory", () => {
    const state = {
      audioPlayer: getInitialAudioPlayerState(),
      itemsByCategory: {
        book: {
          results: ["123"]
        }
      },
      itemsById: {
        123: {
          category: "book",
          item_id: "123",
          link: "link",
          title: "title"
        }
      },
      loading: {}
    };
    const props = {
      navigation: {
        getParam: jest.fn(() => "book"),
        state: {
          params: {
            category: "book"
          }
        }
      } as any
    };
    const getItems = getItemsByCategory();
    const results = getItems(state, props);
    getItems(state, props);
    expect(getItems.recomputations()).toEqual(1);
    expect(results).toMatchSnapshot();
  });
});
