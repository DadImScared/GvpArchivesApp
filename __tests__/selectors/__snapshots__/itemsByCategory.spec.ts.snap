// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`itemsByCategory selector should getItemsByCategory 1`] = `
Object {
  "isLoading": false,
  "nextPage": false,
  "results": Array [
    Object {
      "category": "book",
      "item_id": "123",
      "link": "link",
      "title": "title",
    },
  ],
  "timeToLive": 0,
}
`;
